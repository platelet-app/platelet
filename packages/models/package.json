{
    "name": "@platelet-app/models",
    "version": "1.0.1",
    "description": "DataStore models for the platelet.app project",
    "homepage": "https://platelet.app",
    "bugs": {
        "url": "https://github.com/platelet-app/platelet/issues"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/platelet-app/platelet.git"
    },
    "license": "Apache-2.0",
    "author": "Theo Cranmore <theo@platelet.app>",
    "type": "module",
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "build": "cd ../.. && amplify api gql-compile && amplify codegen models && cd packages/models && cp ../../models/* dist",
        "prepublishOnly": "npm version patch"
    }
}
