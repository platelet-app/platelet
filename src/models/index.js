// @ts-check
import { initSchema } from '@aws-amplify/datastore';
import { schema } from './schema';

const Role = {
  "USER": "USER",
  "COORDINATOR": "COORDINATOR",
  "RIDER": "RIDER",
  "ADMIN": "ADMIN"
};

const CommentVisibility = {
  "EVERYONE": "EVERYONE",
  "ME": "ME"
};

const TaskStatus = {
  "NEW": "NEW",
  "ACTIVE": "ACTIVE",
  "PICKED_UP": "PICKED_UP",
  "DROPPED_OFF": "DROPPED_OFF",
  "CANCELLED": "CANCELLED",
  "REJECTED": "REJECTED",
  "ABANDONED": "ABANDONED",
  "COMPLETED": "COMPLETED"
};

const Priority = {
  "HIGH": "HIGH",
  "MEDIUM": "MEDIUM",
  "LOW": "LOW"
};

const DeliverableTypeIcon = {
  "BUG": "BUG",
  "CHILD": "CHILD",
  "DOCUMENT": "DOCUMENT",
  "EQUIPMENT": "EQUIPMENT",
  "OTHER": "OTHER"
};

const DeliverableUnit = {
  "NONE": "NONE",
  "LITER": "LITER",
  "MILLILITER": "MILLILITER",
  "GRAM": "GRAM",
  "ITEM": "ITEM",
  "BOX": "BOX"
};

const { User, PossibleRiderResponsibilities, RiderResponsibility, Comment, TaskAssignee, Task, Location, Handover, Deliverable, DeliverableType, VehicleAssignment, Vehicle, Tenant, Statistics, SendFeedback, AddressAndContactDetails, S3Object } = initSchema(schema);

export {
  User,
  PossibleRiderResponsibilities,
  RiderResponsibility,
  Comment,
  TaskAssignee,
  Task,
  Location,
  Handover,
  Deliverable,
  DeliverableType,
  VehicleAssignment,
  Vehicle,
  Tenant,
  Role,
  CommentVisibility,
  TaskStatus,
  Priority,
  DeliverableTypeIcon,
  DeliverableUnit,
  Statistics,
  SendFeedback,
  AddressAndContactDetails,
  S3Object
};