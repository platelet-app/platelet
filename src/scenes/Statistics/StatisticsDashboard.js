import React, { useEffect, useState } from "react";
import * as models from "../../models";
import { PaddedPaper } from "../../styles/common";
import TasksStatistics from "./components/TasksStatistics";
import CircularProgress from "@mui/material/CircularProgress";
import { useDispatch, useSelector } from "react-redux";
import { Fade, Stack, Box, Divider } from "@mui/material";
import { getWhoami } from "../../redux/Selectors";
import getStats from "./utilities/getStats";
import moment from "moment";
import { displayErrorNotification } from "../../redux/notifications/NotificationsActions";
import DaysSelection, { Days } from "../../components/DaysSelection";
import CoordinatorPicker from "../../components/CoordinatorPicker";
import UserChip from "../../components/UserChip";

const initialState = {
    common: {
        numCompleted: 0,
        numPickedUp: 0,
        numActive: 0,
        numUnassigned: 0,
        numRejected: 0,
        numCancelled: 0,
        numTasks: 0,
        timeActive: 0,
    },
    priorities: {
        high: 0,
        medium: 0,
        low: 0,
    },
    riders: {},
};

function StatisticsDashboard() {
    const [state, setState] = useState(initialState);
    const [isFetching, setIsFetching] = useState(false);
    const [adminSelectedUser, setAdminSelectedUser] = useState(null);
    const whoami = useSelector(getWhoami);
    const [days, setDays] = useState(3);
    const dispatch = useDispatch();

    const handleDaysChange = (value) => {
        setDays(value);
    };

    const getStatsData = React.useCallback(async () => {
        try {
            setIsFetching(true);
            const newMoment = moment();
            const start = newMoment.toISOString();
            const end = newMoment.subtract(days, "day").toISOString();
            const range = {
                start,
                end,
            };
            setState(
                await getStats(
                    models.Role.COORDINATOR,
                    range,
                    adminSelectedUser ? adminSelectedUser.id : whoami.id
                )
            );
            setIsFetching(false);
        } catch (e) {
            dispatch(displayErrorNotification("Sorry, something went wrong"));
            setState(initialState);
            console.log(e);
        }
    }, [days, whoami.id, dispatch, adminSelectedUser]);

    useEffect(() => getStatsData(), [getStatsData]);
    const isAdmin = whoami?.roles?.includes(models.Role.ADMIN);

    return (
        <PaddedPaper>
            <Stack
                sx={{ paddingBottom: 3 }}
                divider={<Divider />}
                direction="column"
                spacing={2}
            >
                <Stack
                    direction="row"
                    alignItems="center"
                    justifyContent="space-between"
                >
                    <DaysSelection
                        exclude={[Days.TWO_WEEKS]}
                        value={days}
                        onChange={handleDaysChange}
                    />
                    <Fade
                        in={isFetching}
                        style={{
                            transitionDelay: isFetching ? "800ms" : "0ms",
                        }}
                    >
                        <CircularProgress />
                    </Fade>
                </Stack>
                {isAdmin && !adminSelectedUser && (
                    <Box sx={{ maxWidth: 400 }}>
                        <CoordinatorPicker
                            label="Select a coordinator..."
                            onSelect={(v) => setAdminSelectedUser(v)}
                        />
                    </Box>
                )}
                {adminSelectedUser && (
                    <Box>
                        <UserChip
                            user={adminSelectedUser}
                            onDelete={() => setAdminSelectedUser(null)}
                        />
                    </Box>
                )}
            </Stack>
            <TasksStatistics data={state} />
        </PaddedPaper>
    );
}

export default StatisticsDashboard;
