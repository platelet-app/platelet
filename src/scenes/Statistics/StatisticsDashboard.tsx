import React, { useState } from "react";
import * as models from "../../models";
import { PaddedPaper } from "../../styles/common";
import TasksStatistics from "./components/TasksStatistics";
import CircularProgress from "@mui/material/CircularProgress";
import { useDispatch, useSelector } from "react-redux";
import { Fade, Stack, Box, Divider } from "@mui/material";
import { getWhoami } from "../../redux/Selectors";
import getStats from "./utilities/getStats";
import moment from "moment";
import { displayErrorNotification } from "../../redux/notifications/NotificationsActions";
import DaysSelection, { Days } from "../../components/DaysSelection";
import CoordinatorPicker from "../../components/CoordinatorPicker";
import UserChip from "../../components/UserChip";
import useGetTasksGraphql from "../../hooks/useGetTasksGraphql";
import { ModelSortDirection } from "../../API";

function StatisticsDashboard() {
    const [adminSelectedUser, setAdminSelectedUser] =
        useState<models.User | null>(null);
    const whoami = useSelector(getWhoami);
    const [days, setDays] = useState(3);
    const [startDate, setStartDate] = useState(
        moment().subtract(3, "day").toDate()
    );
    const [endDate, setEndDate] = useState(moment().toDate());
    const { state, isFetching, isFetchingMore, isFinished, getNext, error } =
        useGetTasksGraphql(
            100,
            ModelSortDirection.DESC,
            startDate,
            endDate,
            true
        );
    const dispatch = useDispatch();

    React.useEffect(() => {
        if (error) {
            dispatch(displayErrorNotification("Sorry, something went wrong."));
        }
    }, [error, dispatch]);

    const getMore = React.useCallback(async () => {
        if (!isFinished && !isFetchingMore && !isFetching) {
            console.log("Fetching more tasks...");
            await getNext();
        }
    }, [isFinished, getNext, isFetchingMore, isFetching]);

    React.useEffect(() => {
        getMore();
    }, [getMore]);

    const handleDaysChange = (value: number) => {
        setDays(value);
        setStartDate(moment().subtract(value, "day").toDate());
        setEndDate(moment().toDate());
    };

    const isAdmin = whoami?.roles?.includes(models.Role.ADMIN);

    return (
        <PaddedPaper>
            <Stack
                sx={{ paddingBottom: 3 }}
                divider={<Divider />}
                direction="column"
                spacing={2}
            >
                <Stack
                    direction="row"
                    alignItems="center"
                    justifyContent="space-between"
                >
                    <DaysSelection
                        exclude={[Days.TWO_WEEKS]}
                        value={days}
                        onChange={handleDaysChange}
                        disabled={isFetching || isFetchingMore}
                    />
                    <Fade
                        in={isFetching}
                        style={{
                            transitionDelay: isFetching ? "800ms" : "0ms",
                        }}
                    >
                        <CircularProgress />
                    </Fade>
                </Stack>
                {isAdmin && !adminSelectedUser && (
                    <Box sx={{ maxWidth: 400 }}>
                        <CoordinatorPicker
                            label="Select a coordinator..."
                            onSelect={(v) => setAdminSelectedUser(v)}
                        />
                    </Box>
                )}
                {adminSelectedUser && (
                    <Box>
                        <UserChip
                            user={adminSelectedUser}
                            onDelete={() => setAdminSelectedUser(null)}
                        />
                    </Box>
                )}
            </Stack>
            <TasksStatistics
                data={getStats(state, adminSelectedUser?.id || whoami.id)}
            />
        </PaddedPaper>
    );
}

export default StatisticsDashboard;
