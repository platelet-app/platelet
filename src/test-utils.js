import React from "react";
import createSagaMiddleware from "redux-saga";
import * as models from "./models";
import {
    ThemeProvider,
    StyledEngineProvider,
    createTheme,
} from "@mui/material/styles";
import { CssBaseline } from "@mui/material";
import { SnackbarProvider, withSnackbar } from "notistack";
import { Provider, useDispatch, useSelector } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import mediaQuery from "css-mediaquery";
import ReactNotification from "react-notifications-component";
import "react-notifications-component/dist/theme.css";
import AdapterDateFns from "@mui/lab/AdapterDateFns";
import LocalizationProvider from "@mui/lab/LocalizationProvider";
import { render as rtlRender } from "@testing-library/react";
import { initialiseApp } from "./redux/initialise/initialiseActions";
import rootSaga from "./redux/RootSagas";
import { DismissButton } from "./styles/common";
import SnackNotificationButtons from "./components/SnackNotificationButtons";
import { configureStore } from "@reduxjs/toolkit";
import rootReducer from "./redux/Reducers";
import { userRoles } from "./apiConsts";

const testUserModel = new models.User({
    name: "whoami",
    displayName: "Mock User",
    roles: Object.values(userRoles),
    dateOfBirth: null,
    profilePictureThumbnailURL: null,
});

export const testUser = { ...testUserModel, id: "whoami" };

const taskStatus = {
    NEW: "rgba(252, 231, 121, 1)",
    ACTIVE: "cornflowerblue",
    PICKED_UP: "orange",
    DROPPED_OFF: "lightgreen",
    CANCELLED: "blue",
    REJECTED: "grey",
};
const theme = createTheme({
    palette: {
        mode: "light",
        background: {
            default: "rgb(235, 235, 235)",
        },
        taskStatus,
    },
});

export function createMatchMedia(width) {
    return (query) => ({
        matches: mediaQuery.match(query, {
            width,
        }),
        addListener: () => {},
        removeListener: () => {},
    });
}
function getRandomInt(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min) + min); //The maximum is exclusive and the minimum is inclusive
}

export function generateTimes(previous = null, hours = 2) {
    let date;
    if (previous) date = new Date(previous);
    else date = new Date();
    if (!previous) date.setHours(date.getHours() - hours);
    date.setMinutes(date.getMinutes() + getRandomInt(20, 30));
    const timeOfCall = date.toISOString();
    date.setMinutes(date.getMinutes() + getRandomInt(20, 30));
    const timePickedUp = date.toISOString();
    date.setMinutes(date.getMinutes() + getRandomInt(20, 30));
    const timeDroppedOff = date.toISOString();

    return { timeDroppedOff, timePickedUp, timeOfCall };
}

const sagaOptions = {
    onError: (action, error) => {
        console.log("An uncaught exception has occurred in redux-saga:");
        console.log(action);
        if (error) {
            console.log(error.message);
        }
        throw error;
    },
};

function TestApp(props) {
    const dispatch = useDispatch();
    function initialise() {
        return;
        dispatch(initialiseApp());
    }
    React.useEffect(initialise, []);

    const incomingNotification = useSelector((state) => state.notification);

    const snackDismissAction = (key) => (
        <React.Fragment>
            <DismissButton onClick={() => props.closeSnackbar(key)} />
        </React.Fragment>
    );

    function showNotification() {
        if (incomingNotification) {
            const { message, options, restoreCallback, viewLink } =
                incomingNotification;
            options.action = (key) => (
                <SnackNotificationButtons
                    restoreCallback={restoreCallback}
                    viewLink={viewLink}
                    snackKey={key}
                    closeSnackbar={props.closeSnackbar}
                />
            );
            props.enqueueSnackbar(message, options);
        }
    }

    React.useEffect(showNotification, [incomingNotification]);
    return (
        <StyledEngineProvider injectFirst>
            <ThemeProvider theme={theme}>
                <CssBaseline />
                {props.children}
            </ThemeProvider>
        </StyledEngineProvider>
    );
}

const AppSnacked = withSnackbar(TestApp);

const sagaMiddleWare = createSagaMiddleware(sagaOptions);

function render(
    ui,
    {
        preloadedState,
        store = configureStore({
            reducer: rootReducer,
            middleware: [sagaMiddleWare],
            preloadedState: {
                roleView: "ALL",
                awsHubDataStoreEventsReducer: {
                    networkStatus: true,
                    ready: true,
                },
                ...preloadedState,
            },
        }),
        ...renderOptions
    } = {}
) {
    sagaMiddleWare.run(rootSaga);
    function Wrapper(props) {
        return (
            <Provider store={store}>
                <BrowserRouter>
                    <LocalizationProvider dateAdapter={AdapterDateFns}>
                        <ReactNotification />
                        <SnackbarProvider maxSnack={1}>
                            <AppSnacked {...props}>{props.children}</AppSnacked>
                        </SnackbarProvider>
                    </LocalizationProvider>
                </BrowserRouter>
            </Provider>
        );
    }
    return {
        store,
        component: rtlRender(ui, { wrapper: Wrapper, ...renderOptions }),
    };
}
// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
