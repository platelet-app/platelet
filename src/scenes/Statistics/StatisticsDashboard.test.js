import React from "react";
import * as models from "../../models";
import { DataStore } from "aws-amplify/";
import StatisticsDashboard from "./StatisticsDashboard";
import { render, screen } from "../../test-utils";
import { API } from "aws-amplify";
import * as APITypes from "../../API";
import { waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const tenantId = "tenantId";

let preloadedState = {
    tenantId,
};

describe("StatisticsDashboard", () => {
    const isoDate = "2021-11-29T23:24:58.987Z";
    beforeEach(() => {
        jest.useFakeTimers("modern");
        jest.setSystemTime(new Date(isoDate));
    });
    afterEach(() => {
        jest.restoreAllMocks();
        jest.useRealTimers();
    });
    it("displays statistics for the current user", async () => {
        const mockWhoami = await DataStore.save(
            new models.User({
                tenantId,
                displayName: "someDisplayName",
                roles: [models.Role.ADMIN, models.Role.COORDINATOR],
            })
        );
        preloadedState = { ...preloadedState, whoami: { user: mockWhoami } };
        const mockTask = {
            id: "someId",
            timeOfCall: "2021-05-01T00:00:00.000Z",
            riderResponsibility: "someRiderResponsibility",
            priority: APITypes.Priority.HIGH,
            status: APITypes.TaskStatus.COMPLETED,
            createdAt: new Date().toISOString(),
            assignees: {
                items: [
                    {
                        role: models.Role.RIDER,
                        assignee: {
                            id: "someId",
                            displayName: "Rider Name",
                        },
                    },
                    {
                        role: models.Role.COORDINATOR,
                        assignee: {
                            id: mockWhoami.id,
                            displayName: "Coordinator Name",
                        },
                    },
                ],
            },
            _version: 1,
            _deleted: null,
            _lastChangedAt: 1620000000000,
        };
        const notMyMockTask = {
            id: "anotherId",
            timeOfCall: "2021-05-01T00:00:00.000Z",
            riderResponsibility: "someRiderResponsibility",
            priority: APITypes.Priority.HIGH,
            status: APITypes.TaskStatus.COMPLETED,
            createdAt: new Date().toISOString(),
            assignees: {
                items: [
                    {
                        role: models.Role.RIDER,
                        assignee: {
                            id: "someId",
                            displayName: "Not My Rider Name",
                        },
                    },
                    {
                        role: models.Role.COORDINATOR,
                        assignee: {
                            id: "someOtherId",
                            displayName: "Coordinator Name",
                        },
                    },
                ],
            },
            _version: 1,
            _deleted: null,
            _lastChangedAt: 1620000000000,
        };

        const graphqlSpy = jest.spyOn(API, "graphql").mockResolvedValue({
            data: {
                listTasksByTenantId: {
                    items: [mockTask, notMyMockTask],
                    nextToken: null,
                    startedAt: 1620000000000,
                },
            },
        });
        render(<StatisticsDashboard />, { preloadedState });
        expect(await screen.findByText("Rider Name")).toBeInTheDocument();
        expect(screen.queryByText("Coordinator Name")).toBeNull();
        expect(screen.queryByText("Not My Rider Name")).toBeNull();
        expect(graphqlSpy.mock.calls).toMatchSnapshot();
    });
    it("get further pages from graphql", async () => {
        const mockWhoami = await DataStore.save(
            new models.User({
                tenantId,
                displayName: "someDisplayName",
                roles: [models.Role.ADMIN, models.Role.COORDINATOR],
            })
        );
        preloadedState = { ...preloadedState, whoami: { user: mockWhoami } };
        const mockTask = {
            id: "someId",
            timeOfCall: "2021-05-01T00:00:00.000Z",
            riderResponsibility: "someRiderResponsibility",
            priority: APITypes.Priority.HIGH,
            status: APITypes.TaskStatus.COMPLETED,
            createdAt: new Date().toISOString(),
            assignees: {
                items: [
                    {
                        role: models.Role.RIDER,
                        assignee: {
                            id: "someId",
                            displayName: "Rider Name",
                        },
                    },
                    {
                        role: models.Role.COORDINATOR,
                        assignee: {
                            id: mockWhoami.id,
                            displayName: "Coordinator Name",
                        },
                    },
                ],
            },
            _version: 1,
            _deleted: null,
            _lastChangedAt: 1620000000000,
        };
        const mockTaskSecond = {
            id: "anotherId",
            timeOfCall: "2021-05-01T00:00:00.000Z",
            riderResponsibility: "someRiderResponsibility",
            priority: APITypes.Priority.HIGH,
            status: APITypes.TaskStatus.COMPLETED,
            createdAt: new Date().toISOString(),
            assignees: {
                items: [
                    {
                        role: models.Role.RIDER,
                        assignee: {
                            id: "someId",
                            displayName: "Someone Person",
                        },
                    },
                    {
                        role: models.Role.COORDINATOR,
                        assignee: {
                            id: mockWhoami.id,
                            displayName: "Coordinator Name",
                        },
                    },
                ],
            },
            _version: 1,
            _deleted: null,
            _lastChangedAt: 1620000000000,
        };

        const graphqlSpy = jest
            .spyOn(API, "graphql")
            .mockResolvedValueOnce({
                data: {
                    listTasksByTenantId: {
                        items: [mockTask],
                        nextToken: "something",
                        startedAt: 1620000000000,
                    },
                },
            })
            .mockResolvedValue({
                data: {
                    listTasksByTenantId: {
                        items: [mockTaskSecond],
                        nextToken: null,
                        startedAt: 1620000000000,
                    },
                },
            });
        render(<StatisticsDashboard />, { preloadedState });
        expect(await screen.findByText("Rider Name")).toBeInTheDocument();
        expect(await screen.findByText("Someone Person")).toBeInTheDocument();
        expect(graphqlSpy.mock.calls).toMatchSnapshot();
    });
    it("displays statistics for another user as admin", async () => {
        const mockWhoami = await DataStore.save(
            new models.User({
                tenantId,
                displayName: "someDisplayName",
                roles: [models.Role.ADMIN, models.Role.COORDINATOR],
            })
        );
        const anotherCoordinator = await DataStore.save(
            new models.User({
                tenantId,
                displayName: "anotherCoordinator",
                roles: [models.Role.COORDINATOR],
                disabled: 0,
            })
        );
        const querySpy = jest.spyOn(DataStore, "query");
        preloadedState = { ...preloadedState, whoami: { user: mockWhoami } };
        const mockTask = {
            id: "someId",
            timeOfCall: "2021-05-01T00:00:00.000Z",
            riderResponsibility: "someRiderResponsibility",
            priority: APITypes.Priority.HIGH,
            status: APITypes.TaskStatus.COMPLETED,
            createdAt: new Date().toISOString(),
            assignees: {
                items: [
                    {
                        role: models.Role.RIDER,
                        assignee: {
                            id: "someId",
                            displayName: "Rider Name",
                        },
                    },
                    {
                        role: models.Role.COORDINATOR,
                        assignee: {
                            id: mockWhoami.id,
                            displayName: "Coordinator Name",
                        },
                    },
                ],
            },
            _version: 1,
            _deleted: null,
            _lastChangedAt: 1620000000000,
        };
        const notMyMockTask = {
            id: "anotherId",
            timeOfCall: "2021-05-01T00:00:00.000Z",
            riderResponsibility: "someRiderResponsibility",
            priority: APITypes.Priority.HIGH,
            status: APITypes.TaskStatus.COMPLETED,
            createdAt: new Date().toISOString(),
            assignees: {
                items: [
                    {
                        role: models.Role.RIDER,
                        assignee: {
                            id: "someId",
                            displayName: "Not My Rider Name",
                        },
                    },
                    {
                        role: models.Role.COORDINATOR,
                        assignee: {
                            id: anotherCoordinator.id,
                            displayName: anotherCoordinator.displayName,
                        },
                    },
                ],
            },
            _version: 1,
            _deleted: null,
            _lastChangedAt: 1620000000000,
        };

        jest.spyOn(API, "graphql").mockResolvedValue({
            data: {
                listTasksByTenantId: {
                    items: [mockTask, notMyMockTask],
                    nextToken: null,
                    startedAt: 1620000000000,
                },
            },
        });
        render(<StatisticsDashboard />, { preloadedState });
        await waitFor(() => {
            expect(querySpy).toHaveBeenCalledTimes(1);
        });
        userEvent.type(screen.getByRole("combobox"), "anotherCoordinator");
        userEvent.click(await screen.findByText("anotherCoordinator"));
        expect(
            await screen.findByText("Not My Rider Name")
        ).toBeInTheDocument();
        expect(screen.queryByText("Rider Name")).toBeNull();
    });
    it("change the days", async () => {
        const mockWhoami = await DataStore.save(
            new models.User({
                tenantId,
                displayName: "someDisplayName",
                roles: [models.Role.ADMIN, models.Role.COORDINATOR],
            })
        );
        preloadedState = { ...preloadedState, whoami: { user: mockWhoami } };

        const graphqlSpy = jest.spyOn(API, "graphql").mockResolvedValue({
            data: {
                listTasksByTenantId: {
                    items: [],
                    nextToken: null,
                    startedAt: 1620000000000,
                },
            },
        });
        render(<StatisticsDashboard />, { preloadedState });
        const button7Days = screen.getByRole("button", {
            name: "One week",
        });
        expect(button7Days).toHaveAttribute("aria-disabled");
        await waitFor(() => {
            expect(button7Days).not.toHaveAttribute("aria-disabled");
        });
        userEvent.click(button7Days);
        expect(button7Days).toHaveAttribute("aria-disabled");
        await waitFor(() => {
            expect(button7Days).not.toHaveAttribute("aria-disabled");
        });
        expect(graphqlSpy.mock.calls).toMatchSnapshot();
    });
    it("non-admin can't select user", async () => {
        const mockWhoami = await DataStore.save(
            new models.User({
                tenantId,
                displayName: "someDisplayName",
                roles: [models.Role.COORDINATOR],
            })
        );
        preloadedState = { ...preloadedState, whoami: { user: mockWhoami } };

        jest.spyOn(API, "graphql").mockResolvedValue({
            data: {
                listTasksByTenantId: {
                    items: [],
                    nextToken: null,
                    startedAt: 1620000000000,
                },
            },
        });
        render(<StatisticsDashboard />, { preloadedState });
        expect(screen.queryByRole("combobox")).toBeNull();
    });
});
