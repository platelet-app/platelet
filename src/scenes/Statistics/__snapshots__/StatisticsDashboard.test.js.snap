// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsDashboard change the days 1`] = `
Array [
  Array [
    Object {
      "query": "
    query ListTasksByTenantId(
        $filter: ModelTaskFilterInput
        $limit: Int
        $nextToken: String
        $tenantId: ID!
        $sortDirection: ModelSortDirection
        $startDate: String
        $endDate: String
    ) {
        listTasksByTenantId(
            filter: $filter
            limit: $limit
            nextToken: $nextToken
            tenantId: $tenantId
            sortDirection: $sortDirection
            createdAt: { between: [$startDate, $endDate] }
        ) {
            items {
                id
                timeOfCall
                dateCreated
                riderResponsibility
                priority
                status
                createdAt
                assignees {
                    items {
                        _deleted
                        role
                        task {
                            priority
                        }
                        assignee {
                            id
                            displayName
                            profilePicture {
                                key
                            }
                        }
                    }
                }
                pickUpLocation {
                    name
                    ward
                    line1
                    town
                    postcode
                }
                dropOffLocation {
                    ward
                    line1
                    name
                    town
                    postcode
                }
                deliverables {
                    items {
                        count
                        _deleted
                        deliverableType {
                            label
                            icon
                        }
                    }
                }
                comments {
                    items {
                        id
                        _deleted
                        visibility
                        author {
                            id
                        }
                    }
                }
                _version
                _deleted
                _lastChangedAt
            }
            nextToken
            startedAt
        }
    }
",
      "variables": Object {
        "endDate": "2021-11-29T23:24:58.987Z",
        "limit": 100,
        "sortDirection": "DESC",
        "startDate": "2021-11-26T23:24:58.987Z",
        "tenantId": "tenantId",
      },
    },
  ],
  Array [
    Object {
      "query": "
    query ListTasksByTenantId(
        $filter: ModelTaskFilterInput
        $limit: Int
        $nextToken: String
        $tenantId: ID!
        $sortDirection: ModelSortDirection
        $startDate: String
        $endDate: String
    ) {
        listTasksByTenantId(
            filter: $filter
            limit: $limit
            nextToken: $nextToken
            tenantId: $tenantId
            sortDirection: $sortDirection
            createdAt: { between: [$startDate, $endDate] }
        ) {
            items {
                id
                timeOfCall
                dateCreated
                riderResponsibility
                priority
                status
                createdAt
                assignees {
                    items {
                        _deleted
                        role
                        task {
                            priority
                        }
                        assignee {
                            id
                            displayName
                            profilePicture {
                                key
                            }
                        }
                    }
                }
                pickUpLocation {
                    name
                    ward
                    line1
                    town
                    postcode
                }
                dropOffLocation {
                    ward
                    line1
                    name
                    town
                    postcode
                }
                deliverables {
                    items {
                        count
                        _deleted
                        deliverableType {
                            label
                            icon
                        }
                    }
                }
                comments {
                    items {
                        id
                        _deleted
                        visibility
                        author {
                            id
                        }
                    }
                }
                _version
                _deleted
                _lastChangedAt
            }
            nextToken
            startedAt
        }
    }
",
      "variables": Object {
        "endDate": "2021-11-29T23:24:59.387Z",
        "limit": 100,
        "sortDirection": "DESC",
        "startDate": "2021-11-22T23:24:59.387Z",
        "tenantId": "tenantId",
      },
    },
  ],
]
`;

exports[`StatisticsDashboard displays statistics for the current user 1`] = `
Array [
  Array [
    Object {
      "query": "
    query ListTasksByTenantId(
        $filter: ModelTaskFilterInput
        $limit: Int
        $nextToken: String
        $tenantId: ID!
        $sortDirection: ModelSortDirection
        $startDate: String
        $endDate: String
    ) {
        listTasksByTenantId(
            filter: $filter
            limit: $limit
            nextToken: $nextToken
            tenantId: $tenantId
            sortDirection: $sortDirection
            createdAt: { between: [$startDate, $endDate] }
        ) {
            items {
                id
                timeOfCall
                dateCreated
                riderResponsibility
                priority
                status
                createdAt
                assignees {
                    items {
                        _deleted
                        role
                        task {
                            priority
                        }
                        assignee {
                            id
                            displayName
                            profilePicture {
                                key
                            }
                        }
                    }
                }
                pickUpLocation {
                    name
                    ward
                    line1
                    town
                    postcode
                }
                dropOffLocation {
                    ward
                    line1
                    name
                    town
                    postcode
                }
                deliverables {
                    items {
                        count
                        _deleted
                        deliverableType {
                            label
                            icon
                        }
                    }
                }
                comments {
                    items {
                        id
                        _deleted
                        visibility
                        author {
                            id
                        }
                    }
                }
                _version
                _deleted
                _lastChangedAt
            }
            nextToken
            startedAt
        }
    }
",
      "variables": Object {
        "endDate": "2021-11-29T23:24:58.987Z",
        "limit": 100,
        "sortDirection": "DESC",
        "startDate": "2021-11-26T23:24:58.987Z",
        "tenantId": "tenantId",
      },
    },
  ],
]
`;

exports[`StatisticsDashboard get further pages from graphql 1`] = `
Array [
  Array [
    Object {
      "query": "
    query ListTasksByTenantId(
        $filter: ModelTaskFilterInput
        $limit: Int
        $nextToken: String
        $tenantId: ID!
        $sortDirection: ModelSortDirection
        $startDate: String
        $endDate: String
    ) {
        listTasksByTenantId(
            filter: $filter
            limit: $limit
            nextToken: $nextToken
            tenantId: $tenantId
            sortDirection: $sortDirection
            createdAt: { between: [$startDate, $endDate] }
        ) {
            items {
                id
                timeOfCall
                dateCreated
                riderResponsibility
                priority
                status
                createdAt
                assignees {
                    items {
                        _deleted
                        role
                        task {
                            priority
                        }
                        assignee {
                            id
                            displayName
                            profilePicture {
                                key
                            }
                        }
                    }
                }
                pickUpLocation {
                    name
                    ward
                    line1
                    town
                    postcode
                }
                dropOffLocation {
                    ward
                    line1
                    name
                    town
                    postcode
                }
                deliverables {
                    items {
                        count
                        _deleted
                        deliverableType {
                            label
                            icon
                        }
                    }
                }
                comments {
                    items {
                        id
                        _deleted
                        visibility
                        author {
                            id
                        }
                    }
                }
                _version
                _deleted
                _lastChangedAt
            }
            nextToken
            startedAt
        }
    }
",
      "variables": Object {
        "endDate": "2021-11-29T23:24:58.987Z",
        "limit": 100,
        "sortDirection": "DESC",
        "startDate": "2021-11-26T23:24:58.987Z",
        "tenantId": "tenantId",
      },
    },
  ],
  Array [
    Object {
      "query": "
    query ListTasksByTenantId(
        $filter: ModelTaskFilterInput
        $limit: Int
        $nextToken: String
        $tenantId: ID!
        $sortDirection: ModelSortDirection
        $startDate: String
        $endDate: String
    ) {
        listTasksByTenantId(
            filter: $filter
            limit: $limit
            nextToken: $nextToken
            tenantId: $tenantId
            sortDirection: $sortDirection
            createdAt: { between: [$startDate, $endDate] }
        ) {
            items {
                id
                timeOfCall
                dateCreated
                riderResponsibility
                priority
                status
                createdAt
                assignees {
                    items {
                        _deleted
                        role
                        task {
                            priority
                        }
                        assignee {
                            id
                            displayName
                            profilePicture {
                                key
                            }
                        }
                    }
                }
                pickUpLocation {
                    name
                    ward
                    line1
                    town
                    postcode
                }
                dropOffLocation {
                    ward
                    line1
                    name
                    town
                    postcode
                }
                deliverables {
                    items {
                        count
                        _deleted
                        deliverableType {
                            label
                            icon
                        }
                    }
                }
                comments {
                    items {
                        id
                        _deleted
                        visibility
                        author {
                            id
                        }
                    }
                }
                _version
                _deleted
                _lastChangedAt
            }
            nextToken
            startedAt
        }
    }
",
      "variables": Object {
        "endDate": "2021-11-30T00:00:00.000Z",
        "limit": 100,
        "nextToken": "something",
        "sortDirection": "DESC",
        "startDate": "2021-11-26T00:00:00.000Z",
        "tenantId": "tenantId",
      },
    },
  ],
]
`;
